index.d.ts

```diff
@@ -0,0 +1,70 @@
+import { ComponentClass, Component, StatelessComponent, ReactNode } from 'react';
+import { Store, Dispatch, ActionCreator } from 'redux';
+
+interface ComponentDecorator<TOriginalProps, TOwnProps> {
+  (component: ComponentClass<TOriginalProps>|StatelessComponent<TOriginalProps>): ComponentClass<TOwnProps>;
+}
+
+/**
+ * Following 3 functions cover all possible ways connect could be invoked
+ *
+ * - State: Redux state interface (the same one used by Store<S>)
+ * - TStateProps: Result of MapStateToProps
+ * - TDispatchProps: Result of MapDispatchToProps
+ * - TOwnProps: Props passed to the wrapping component
+ */
+export function connect<State, TOwnProps>(): ComponentDecorator<{ dispatch: Dispatch<State> }, TOwnProps>;
+
+export function connect<State, TOwnProps, TStateProps>(
+  mapStateToProps: MapStateToProps<State, TStateProps, TOwnProps>,
+): ComponentDecorator<TStateProps & { dispatch: Dispatch<State> }, TOwnProps>;
```

Comment:
Same as previous one. ComponentDecorator exposes the exported component with `TOwnProps` outside, with `TStateProps & { dispatch: Dispatch<State> }` satisfied by React-Redux.
I think you are possibly misreading what `ComponentDecorator` does or how which props of the HOC should be exposed outside.
I've tested that part extensively in my (big) app and it really behaves as supposed to.